src/search.cpp

```diff
@@ -566,8 +567,9 @@ Value Search::Worker::search(
         // Step 2. Check for aborted search and immediate draw
         if (threads.stop.load(std::memory_order_relaxed) || pos.is_draw(ss->ply)
             || ss->ply >= MAX_PLY)
-            return (ss->ply >= MAX_PLY && !ss->inCheck) ? evaluate(pos, thisThread->optimism[us])
-                                                        : value_draw(thisThread->nodes);
+            return (ss->ply >= MAX_PLY && !ss->inCheck)
+                   ? evaluate(networks, pos, thisThread->optimism[us])
```

Comment:
I feel like passing `networks` to the evaluate function in search every time is quite counterproductive. Perhaps the function call in search could be kept the same?

Reply:
how else would evaluate() access the networks?

Reply:
@Disservin @Sopel97 
Perhaps it would be more natural for the Eval namespace to store a reference to the networks rather than the Worker class storing it?  It seems like search really should not know anything about any networks.

Reply:
@mstembera ideally what we would like to achieve is to make it possible to create multiple stockfish's so that we have proper isolation for #4626, when you move it into the eval namespace I think this is going a step back again? Though I agree that it is perhaps a bit weird for search to know about the networks.. ideally search should get an "evaluator" which handles evaluation I guess?

Reply:
search knowing networks is nothing wrong, since logically it uses them as a parameter of evaluation

Reply:
@Disservin I think making SF a library is a nice goal and I don't want to get in the way of that.  In either case though there is only one actual instance of the networks and the rest are just references to it whether in Eval or Worker so it should not be a step backwards?   Maybe I'm missing something.

@Sopel97 I didn't mean to imply that it is wrong but I think less dependencies if they are unnecessary is better.

Maybe if I can make it work I will open a RFC PR so we can discuss more concretely.

Reply:
> Maybe if I can make it work I will open a RFC PR so we can discuss more concretely.

Sure!

> Maybe I'm missing something.

What if the user instantiates two stockfish engines, with different network weights, I don't see how that would exactly work with your approach.

Reply:
Ah ok I didn't know we also wanted to support different network weights for different instances.  In that case I think your idea of turning the Eval namespace into and Evaluator class would solve it.

Reply:
Yeah I think overall it'd be good to have.. on the weekend i quickly spun something to have uci and the engine a bit more split up https://github.com/official-stockfish/Stockfish/compare/master...Disservin:Stockfish:uci-split-engine, I imagine something in that direction, though that version isnt particularly useful right now for a library since everything returns void. Though I think it already looks quite good in terms of how "straight forward" the uci implementation is. 

If you want to propose something for the evaluator class feel free to open a draft or the like.